### PR TITLE
Add the ability to customise the CocoaPods-Try via a .cocoapods.yml

### PR DESCRIPTION
**Pull request by orta**
Adds the ability to run custom pre installation steps for `pod try` via a `.cocoapods.yml`. Should allow for fixing #38 and #33

### CODE REVIEW

File: spec/command/try_settings_spec.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        settings.should.be.instance_of TrySettings
+        settings.pre_install_commands.should.be.nil?
+        settings.project_path.should.be.nil?
+      end
+    end
+
+    it 'returns an instance with the right defaults when there are no yml settings files' do
+      Dir.mktmpdir do |dir|
+        yaml = <<eos
+          try_pre_install:
```
Comment (kylef) [inline spec/command/try_settings_spec.rb]: Why not use something like this:

```yaml
try:
  install:
    pre:
      - pod install
      - git submodule init
  workspace: ORStackView.xcworkspace
```

That way all plugin related things are grouped at the root.

Comment (orta) [inline spec/command/try_settings_spec.rb]: I like this idea :+1:

**Comment by orta** [comment]
Alright, I've also updated the `Gemfile.lock` and removed one of the new checks in RuboCop - it was only applying to the tests.

Ran it with against a fork of a pod I had laying around. `bundle exec pod try https://github.com/orta/AMTagListView --verbose` Works :+1:

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,26 @@
 # Cocoapods::Try Changelog
 
+## Master
+
+##### Enhancements
+
+* A CocoaPods can now declare their own pre-install commands, and prefer a project.
+  To use this, add a `.cocoapods.yml` file to your repo. The yaml file should have
```
Comment (kylef) [inline CHANGELOG.md]: file in the **root** of your repository.?

Comment (orta) [inline CHANGELOG.md]: :+1:

File: lib/pod/try_settings.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      try_settings = TrySettings.new
+      try_settings.pre_install_commands = Array(settings['try']['install']['pre'])
+      if settings['try']['project']
+        try_settings.project_path = Pathname.new(path) + settings['try']['project']
+      end
+      try_settings
+    end
+
+    # If we need to run commands from pod-try we should let the users know
+    # what is going to be running on their device.
```
Comment (kylef) [inline lib/pod/try_settings.rb]: :+1: awesome!

File: lib/pod/try_settings.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    # Creates a TrySettings instance based on a folder path
+    #
+    def self.settings_from_folder(path)
+      settings_path = Pathname.new(path) + '.cocoapods.yml'
+      return TrySettings.new unless File.exist? settings_path
+
+      settings = YAML.load(File.read(settings_path))
+      try_settings = TrySettings.new
+      try_settings.pre_install_commands = Array(settings['try']['install']['pre'])
```
Comment (kylef) [inline lib/pod/try_settings.rb]: Wouldn't this crash if we have a settings file and it doesn't have an install pre setting?

Comment (orta) [inline lib/pod/try_settings.rb]:
```
  - handles not including system commands in the pre-install [FAILED]
  - handles not including the try in the yaml [FAILED]
```

Yep, great catch @kylef - thanks. Will fix,

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,26 @@
 # Cocoapods::Try Changelog
 
+## Master
+
+##### Enhancements
+
+* A CocoaPods can now declare their own pre-install commands, and prefer a project.
```
Comment (alexito4) [inline CHANGELOG.md]: Shouldn't this say "A Pod..."? :question:

Comment (orta) [inline CHANGELOG.md]: Any, and any repo too, changed it.

File: .rubocop.yml [hunk 6]
```diff
@@ -1,3 +1,6 @@
 inherit_from:
  - rubocop-todo.yml
  - .rubocop_cocoapods.yml
+
+Metrics/ModuleLength:
+  Enabled: false
```
Comment (segiddins) [inline .rubocop.yml]: iirc this is already disabled in the standard CocoaPods config?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,26 @@
...
+  try:
+    install:
+      pre:
+        - pod install
+        - git submodule init
+    project: 'ORStackView.xcworkspace'
+  ```
+
+  [Orta Therox](https://github.com/orta)
+  Fixes [cocoapods-try#33](https://github.com/CocoaPods/cocoapods-try/issues/33)
```
Comment (segiddins) [inline CHANGELOG.md]: we normally only link to the issue, and don't prefix the number when its from the same repo

File: lib/pod/try_settings.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    # If we need to run commands from pod-try we should let the users know
+    # what is going to be running on their device.
+    #
+    def prompt_for_permission
+      UI.titled_section 'Running Pre-Install Commands' do
+        commands = pre_install_commands.length > 1 ? 'commands' : 'command'
+        UI.puts "In order to try this pod, CocoaPods-Try needs to run the following #{commands}:"
+        pre_install_commands.each { |command| UI.puts " - #{command}" }
+        UI.puts "\nPress return to run these #{commands}, or press `cmd + c` to stop trying this pod."
```
Comment (segiddins) [inline lib/pod/try_settings.rb]: control-c sends SIGINT

File: lib/pod/try_settings.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    # Runs the pre_install_commands from
+    #
+    # @param  [Bool] prompt
+    #         Should CocoaPods-Try show a prompt with the commands to the user.
+    #
+    def run_pre_install_commands(prompt)
+      if pre_install_commands
+        prompt_for_permission if prompt
+        pre_install_commands.each { |command| system command }
```
Comment (segiddins) [inline lib/pod/try_settings.rb]: Let's do what we do for the prepare command here?

```ruby
Executable.execute_command('bash', ['-e', command], true)
```

**Comment by segiddins** [comment]
Excellent work @orta ! Awesome test coverage :)

File: lib/pod/try_settings.rb (hunk 8 above)
Comment (orta) [inline lib/pod/try_settings.rb]: can we capture this and elegantly end?

File: .rubocop.yml (hunk 6 above)
Comment (orta) [inline .rubocop.yml]: Not in try, I can pull in  anew one from CP ( I updated the rubocop gem )

File: lib/pod/try_settings.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  class TrySettings
+    attr_accessor :pre_install_commands, :project_path
+
+    # Creates a TrySettings instance based on a folder path
+    #
+    def self.settings_from_folder(path)
+      settings_path = Pathname.new(path) + '.cocoapods.yml'
+      return TrySettings.new unless File.exist? settings_path
+
+      settings = YAML.load(File.read(settings_path))
```
Comment (segiddins) [inline lib/pod/try_settings.rb]: use `YAMLHelper.load_file(settings_path)`

File: lib/pod/try_settings.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+      try_settings = TrySettings.new
+      return try_settings unless settings['try']
+
+      if settings['try']['install']
```
Comment (segiddins) [inline lib/pod/try_settings.rb]: no need for the `if` -- `Array(nil)` returns `[]`

File: lib/pod/try_settings.rb (hunk 8 above)
Comment (segiddins) [inline lib/pod/try_settings.rb]: hm? I mean it's control-c, not command-c.

File: spec/command/try_settings_spec.rb [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        settings = TrySettings.settings_from_folder dir
+        settings.should.be.instance_of TrySettings
+        settings.pre_install_commands.should.be.nil?
+        settings.project_path.should.be.nil?
+      end
+    end
+
+    it 'returns an instance with the right defaults when there are no yml settings files' do
+      Dir.mktmpdir do |dir|
+        yaml = <<eos
```
Comment (segiddins) [inline spec/command/try_settings_spec.rb]: `YAML` instead of `eos` so we can get syntax highlighting?

**Comment by segiddins** [comment]
:+1: feel free to merge once the comments are addressed.

File: lib/pod/try_settings.rb (hunk 8 above)
Comment (orta) [inline lib/pod/try_settings.rb]: :D

File: lib/pod/try_settings.rb (hunk 11 above)
Comment (orta) [inline lib/pod/try_settings.rb]: but aren't these dicts?

**Comment by segiddins** [comment]
:shipit:

**Comment by orta** [comment]
:tada:

**Pull request merged**